experimental/python/wheel.bzl

```diff
@@ -0,0 +1,142 @@
+    outfile = ctx.actions.declare_file('-'.join([
+        ctx.attr.wheel_name,
+        ctx.attr.version,
+        ctx.attr.python,
+        ctx.attr.abi,
+        ctx.attr.platform
+    ]) + '.whl')
+
+    inputs = depset(
+        transitive=[dep[DefaultInfo].data_runfiles.files for dep in
+                    ctx.attr.deps] +
+                   [dep[DefaultInfo].default_runfiles.files for dep in
+                    ctx.attr.deps]
+    )
+    arguments = ['--name', ctx.attr.wheel_name,
+                 '--version', ctx.attr.version,
+                 '--python', ctx.attr.python,
+                 '--abi', ctx.attr.abi,
+                 '--platform', ctx.attr.platform,
+                 '--out', outfile.path]
+    arguments.extend(['--restrict_package=%s' % p for p in ctx.attr.packages])
+    for input_file in inputs.to_list():
+        arguments.append('--input_file')
+        arguments.append('%s;%s' % (input_file.short_path, input_file.path))
+
+    extra_headers = []
+    if ctx.attr.author:
+        extra_headers.append("Author: %s" % ctx.attr.author)
+    if ctx.attr.author_email:
+        extra_headers.append("Author-email: %s" % ctx.attr.author_email)
+    if ctx.attr.homepage:
+        extra_headers.append("Home-page: %s" % ctx.attr.homepage)
+    if ctx.attr.license:
+        extra_headers.append("License: %s" % ctx.attr.license)
+
+    for h in extra_headers:
+        arguments.append('--header')
+        arguments.append(h)
+
+    for c in ctx.attr.classifiers:
+        arguments.append('--classifier')
+        arguments.append(c)
+
+    for r in ctx.attr.requires:
+        arguments.append('--requires')
+        arguments.append(r)
+
+    for option, requirements in ctx.attr.extra_requires.items():
+        for r in requirements:
+            arguments.append('--extra_requires')
+            arguments.append(r + ';' + option)
+
+    for name, ref in ctx.attr.console_scripts.items():
+        arguments.append('--console_script')
+        arguments.append(name + ' = ' + ref)
+
+    if ctx.attr.description_file:
+        description_files = ctx.attr.description_file.files.to_list()
+        arguments.append('--description_file')
+        arguments.append(description_files[0].path)
+        inputs = inputs.union(ctx.attr.description_file.files)
+
+    ctx.actions.run(
+        inputs=inputs,
+        outputs=[outfile],
+        arguments=arguments,
+        executable=ctx.executable._wheelmaker,
+        progress_message="Building wheel")
+    return [DefaultInfo(files=depset([outfile]),
+                        data_runfiles=ctx.runfiles(files=[outfile]))]
+
+
+py_wheel = rule(
+    implementation=_py_wheel_impl,
+    attrs={
+        "deps": attr.label_list(
+            doc="Dependencies of this package, e.g. py_library rules"),
+        "packages": attr.string_list(
+            mandatory=True,
+            allow_empty=False,
+            doc="""\
+List of Python packages to include in the distribution.
```

Comment:
This is controversial and perhaps worth discussing.

A typical package will have dependencies on other packages; some might be first party but packaged separately, some might be third party. From bazel point of view they will probably form a tree of py_library dependencies.

When creating a wheel package one does not want to include the whole tree, so we need to tell bazel where to cut the chain. I see three ways to do that:
1) Don't follow dependencies. Only package up the libraries listed directly in deps.
2) Follow dependencies and only package specific python packages
  a) Including sub-packages, so e.g. if "foo" is listed then "foo.bar" gets packaged too
  b) Not including sub-packages.

The implementation here does "2a", but I could see how other options could be preferred. I'm happy to change if you think other options are better.